crates/breez-sdk/common/src/sync/background.rs

```diff
@@ -0,0 +1,1365 @@
+                _ = shutdown_receiver.changed() => {
+                    debug!("Shutdown signal received, stopping update subscription");
+                    break;
+                }
+
+                maybe_notification = stream.message() => {
+                    match maybe_notification {
+                        Ok(Some(notification)) => {
+                            debug!("Received notification for client id: {:?}", notification.client_id);
+                            if let Some(client_id) = notification.client_id && client_id == self.client.client_id {
+                                debug!("Ignoring notification for ourselves");
+                                continue;
+                            }
+
+                            if let Err(e) = tx.send(()) {
+                                error!("Failed to send update notification: {}", e);
+                                break;
+                            }
+                        }
+                        Ok(None) => {
+                            debug!("Notification stream closed by server");
+                            break;
+                        }
+                        Err(e) => {
+                            error!("Error receiving notification: {}", e);
+                            break;
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    /// The main sync loop. Runs in an event loop to ensure no updates happen in parallel.
+    fn start_sync_loop_task(
+        self: &Arc<Self>,
+        shutdown_receiver: watch::Receiver<()>,
+        pull_trigger: watch::Receiver<()>,
+    ) {
+        let clone = Arc::clone(self);
+        tokio::spawn(async move { clone.sync_loop(shutdown_receiver, pull_trigger).await });
+    }
+
+    async fn sync_loop(
+        &self,
+        mut shutdown_receiver: watch::Receiver<()>,
+        mut pull_trigger: watch::Receiver<()>,
+    ) {
+        let mut push_trigger = self.push_sync_trigger.resubscribe();
+        let (backoff_trigger_tx, mut backoff_trigger_rx) = mpsc::channel::<Duration>(10);
+
+        // Mutex to ensure there is only one backoff running at a time.
+        let backoff_handle: Mutex<Option<BackoffHandle>> = Mutex::new(None);
+
+        loop {
+            let (incoming_count, outgoing_count) = tokio::select! {
+                _ = shutdown_receiver.changed() => {
+                    debug!("Shutdown signal received, stopping push sync loop");
+                    break;
+                }
+                _ = pull_trigger.changed() => {
+                    debug!("Received incoming sync notification");
+                    match self.pull_sync_once().await {
+                        Ok(count) => (count, None),
+                        Err(e) => {
+                            error!("Failed to sync once: {}", e);
+                            (None, None)
+                        }
+                    }
+                }
```

Comment:
Awaiting futures within `select` arms prevents those futures from being canceled by the shutdown receiver. In the past, we saw issues with that on Wasm Node.js (resulting in memory leaks). A simple way to avoid it is to put all logic in the async expression and avoid it in the handler.

This also applies below

Reply:
Let's leave as is. We can fix it in another PR. There are likely other places in the codebase where similar issues are prone and we can apply a common solution like we did in the liquid sdk: https://github.com/breez/breez-sdk-liquid/blob/973fd2327a77ef1bceaec4502aeace8e053ec358/lib/core/src/utils/mod.rs#L196